File: services/ingestion-openmeteo/app/Application/Outbox/Contracts/OutboxRepository.php
<?php

namespace App\Application\Outbox\Contracts;

use App\Application\Outbox\DTO\OutboxMessageData;

/**
 * Repository interfejs za rad sa outbox porukama.
 *
 * Application sloj ne treba da zna kako se outbox poruke čuvaju,
 * već samo da može da ih sačuva i kasnije da ih publisher preuzme.
 */
interface OutboxRepository
{
    /**
     * Čuva jednu ili više outbox poruka.
     *
     * @param OutboxMessageData[] $messages
     */
    public function storeMany(array $messages): void;

    /**
     * Uzimа pending poruke koje su spremne za publish i zaključava ih za trenutni proces.
     *
     * @return array<int, array<string, mixed>>
     */
    public function lockPendingBatch(int $limit): array;

    /**
     * Markira poruku kao uspešno poslatu.
     */
    public function markAsPublished(int $id): void;

    /**
     * Markira poruku kao neuspešnu i zakazuje sledeći pokušaj.
     */
    public function markAsFailed(int $id, string $error, int $attempts, int $maxAttempts): void;
}

File: services/ingestion-weatherapi/internal/application/outbox/contracts/outbox_repository.go
package contracts

import (
	"context"

	outboxdto "ingestion-weatherapi/internal/application/outbox/dto"
)

/*
OutboxRepository predstavlja application contract za rad sa outbox porukama.
*/
type OutboxRepository interface {
	/*
		StoreMany čuva jednu ili više outbox poruka.
	*/
	StoreMany(ctx context.Context, messages []outboxdto.OutboxMessageData) error

	/*
		LockPendingBatch zaključava batch pending poruka za publish.

		FOR UPDATE SKIP LOCKED nam omogućava da više publisher procesa
		ne uzmu iste poruke u isto vreme.
	*/
	LockPendingBatch(ctx context.Context, limit int) ([]outboxdto.OutboxMessageData, error)

	/*
		MarkAsPublished označava poruku kao uspešno objavljenu.
	*/
	MarkAsPublished(ctx context.Context, id int64) error

	/*
		MarkAsFailed označava poruku kao failed ili je vraća u pending sa backoff-om.
	*/
	MarkAsFailed(ctx context.Context, id int64, errorMessage string, attempts int, maxAttempts int) error
}
